Modernize / Modularize xacro

rhaschke opened the pull request:
This is the new PR replacing #109 / #113 already announced in https://github.com/shadow-robot/sr_common/pull/113#pullrequestreview-768217742.

### Summary
The key goal followed by the restructuring is a simple and scalable approach to specify different hand variants.
Until, for each variant essentially a separate xacro file was required. With this PR, we can essentially resolve everything using either `full_hand.urdf.xacro` or `hand_lite.urdf.xacro`. Using the macro parameter `fingers` one can specify which fingers are needed. With parameters `[tip|mid|prox]_sensor(s)` one can choose individual (tactile) sensors. The already existing variable `hand_type` can now also be used to specify `muscle` or `muscletrans` hands.

### Changes w.r.t. #109 / #113:
- Removed redundant definition of pi, causing warnings: `redefining global symbol: pi`
- Removed unused meshes in `ubi_tactiles`
- Factored out #118 and #119
- Added a compatibility-test to ensure that the old robot xacros produce the same results as the new ones.
  The test introduced in #109 was just ensuring that the xacros can be processed at all.

### TODOs:
- [x] Approval from @guihomework: https://github.com/shadow-robot/sr_common/pull/120#issuecomment-933397546
- [x] Fix the remaining problem of some lost `xmlns` attributes (see https://github.com/shadow-robot/sr_common/pull/120#issuecomment-932248944).
Thanks to @guihomework [analysis](https://github.com/shadow-robot/sr_common/pull/120#issuecomment-932613152), I was able to fix the remaining problem with a fix in xacro:
   - [x] Merge https://github.com/ros/xacro/pull/287
- [x] Drop commits marked as `TEMP`. They just illustrate the progress of the compatibility test.
- [x] Squash some commits (d891721, 51bf390, ceacb17) into the main one (929fac3). I kept them to illustrate my changes.
- [x] Squash the fixing commits
  I kept the original, unsquashed version of this PR in https://github.com/ubi-agni/sr_common/tree/PR%23120-long.
  One can easily validate the backward compatibility there by inspecting the [history of `test/robots.urdf`](https://github.com/ubi-agni/sr_common/commits/PR%23120-long/sr_description/test/robots.urdf)
- [x] Merge https://github.com/ubi-agni/sr_common/pull/1
   - [x] Merge https://github.com/ros/xacro/pull/283
- [ ] Merge https://github.com/ubi-agni/sr_common/pull/2 ?
- [x] Merge https://github.com/ubi-agni/sr_common/pull/3
   - [x] Merge https://github.com/ros/xacro/pull/284
- [x] To further improve readability, I suggest replacing the list variables with dotified dicts, e.g. `tip_sensors[2]` becomes `tip_sensor.mf`. Argument would be passed as `tip_sensors=${dict(th='bio', mf='std', rf='eli')}`. 
To simplify for the common case of identical values, we could also think of some internal property re-assignment to allow for `tip_sensors=bio`.
- [ ] @beatrizleon decide and @rhaschke? implement the `hand_type` naming: https://github.com/shadow-robot/sr_common/pull/120#issuecomment-933005809
- [ ] Release xacro
  - [x] Melodic: https://github.com/ros/rosdistro/pull/30921
  - [x] Noetic: https://github.com/ros/rosdistro/pull/30922
  - [ ] Sync planned for Oct 13th: https://discourse.ros.org/t/preparing-for-noetic-sync-2021-09-23/22378/5

rhaschke committed:
Replace $(find xacro)/xacro -> xacro

rhaschke committed:
Remove deprecated --inorder option

rhaschke committed:
Fix missing xacro namespace prefixes

guihomework committed:
Fix typo

guihomework committed:
Improved xacro for modularity and to permit sensor selection

Modern xacro allows to automatically forward macro arguments when they are
defined with the arg:=^ syntax. Using this feature simplifies macro calling.

Also, instead of specifying hand attributes with an ever-growing set of
binary parameters, this commit introduces parameters hand_type and
(palm|prox|mid|tip)_sensor to modularize parameterization.
This allows for more flexible definition of the (tactile) sensors used across the hand.

Using a value of 'none' for those sensors allows to deactivate (Gazebo) sensors.

Backward compatibility for old parameters is maintained via file backward_compat.urdf.xacro,
which translates from old to new parameters.

rhaschke committed:
Remove redundant definition of pi

pi is available globally in xacro as python's math.pi

rhaschke committed:
Add compatibility test

Cloned all files from robots/*.urdf.xacro prior to this PR into test/robots.old.
Run xacro on all corresponding files and compare their content.

rhaschke committed:
Fix backward compatibility

- Restore deleted files required by compatibility test
- Fix use of deprecated properties
- Fix use of deprecated parameters
- Undo unrelated changes w.r.t. upstream
- Fix extra_lite hands: it is missing MF not FF

rhaschke committed:
Delete unused meshes in ubi_tactiles

rhaschke commented:
@beatrizleon: How can I access the results of the failing CI jobs at AWS?

rhaschke committed:
Issue deprecation warning

Use python's None value as default to know whether old parameters were specified.

rhaschke committed:
More simplifications

- Simpify one_finger_unit.urdf.xacro
  There is only one finger. No need to pass *_sensors. A single one is sufficient.
  Also, a fingers parameter doesn't make sense.

- Reduce number of xacro:if in fingers.urdf.xacro

- Plain string default parameters don't need to be quoted

rhaschke commented:
I added a series of commits fixing most compatibility-test issues. The only remaining issue are some lost `xmlns` attributes as illustrated in [this comparison](https://github.com/shadow-robot/sr_common/pull/120/commits/0d69acb909105540037d5ec77f72854ba25396e3) showing the differences between generated URDF files from [`test/robots.old`](https://github.com/ubi-agni/sr_common/tree/pr-modernize-modularize-xacro/sr_description/test/robots.old) vs. [`robots`](https://github.com/ubi-agni/sr_common/tree/pr-modernize-modularize-xacro/sr_description/robots). That's what the [compatibility test](https://github.com/ubi-agni/sr_common/blob/pr-modernize-modularize-xacro/sr_description/test/test_compatibility.py) compares as well. I'm not sure where these attributes got lost.
@guihomework, maybe you have an idea?

The good news is that [processing the new robot URDFs yields the same results as in](https://github.com/ubi-agni/sr_common/compare/a741b66751fad674908004670e4404544fca8eaa...pr-modernize-modularize-xacro#diff-b3514a5fc89a4e3ebad511ac51b604aa505f9b0f058850c266ce48c202141e3d) a741b66.
That's already much better than the results from the original PR, shown in [this and following diffs](https://github.com/ubi-agni/sr_common/compare/a741b66751fad674908004670e4404544fca8eaa...515e5e8440d226bb933eb43ed09244846ff94591#diff-dcac675218bab0bff9afd23e9bb6a6e966772c3688e60b4f939dba4342ffd764).

rhaschke committed:
More specific variable tests

Using `'text' in variable` is a rather sloppy equivalence check.
This commit attempts to harden these checks using the following regexp
replacements:

- More specific test of hand_type
  ('hand.+?') in ([a-z_]+)  ->  $2.startswith($1)

- More specific variable tests
  ('.+') in ([a-z_]+)  ->  $2 == $1

rhaschke committed:
More readable hand_types: muscle, muscletrans, motor, motor+

- hand_mt -> muscletrans
- hand_m -> muscle, regexp: (["' ])hand_m -> $1muscle
- hand_e_plus -> motor+
- hand_e -> motor

rhaschke committed:
Simplify syntax for sensor specification

Use textual sensor specifications:
- tip_sensors=bio
- tip_sensors='th=bio, ff=bio, mf=std, rf=std, lf=none'
- tip_sensors='bio, bio, std, std, none'

- Rename backward_compat.urdf.xacro -> process_sensor_parameters.xacro
- Introduce macro validate_tokens

guihomework commented on sr_description/hand/xacro/finger/biotac/biotac.urdf.xacro:
```diff
@@ -7,13 +7,12 @@ xmlns:body="http://playerstage.sourceforge.net/gazebo/xmlschema/#body"
   <!-- parent is the parent link for the joint -->
   <xacro:macro name="biotac_middle"
-  params="muscle prefix link_prefix joint_prefix parent">
+  params="prefix:=^ link_prefix:=^ joint_prefix:=^ tip_sensor:=^|'std' mid_sensor:=^|standard parent">
```
```suggestion
  params="prefix:=^ link_prefix:=^ joint_prefix:=^ tip_sensor:=^|'std' mid_sensor:=^|'std' parent">
```
seems I forgot this one

guihomework commented on sr_description/hand/xacro/finger/distal/distal.gazebo.xacro:
```diff
@@ -1,28 +1,28 @@
-          <bumperTopicName>contacts/${prefix}${link_prefix}/distal</bumperTopicName>
-        </plugin>
-      </sensor>
+  <xacro:macro name="distal_gazebo" params="prefix:=^ link_prefix:=^ tip_sensor:=^|std ns:=^|''">
```
```suggestion
  <xacro:macro name="distal_gazebo" params="prefix:=^ link_prefix:=^ tip_sensor:=^|'std' ns:=^|''">
```

guihomework commented on sr_description/hand/xacro/finger/proximal/proximal.gazebo.xacro:
```diff
@@ -1,24 +1,24 @@
-        </plugin>
-      </sensor>
-        
+  <xacro:macro name="proximal_gazebo" params="prefix:=^ link_prefix:=^ prox_sensor:=^|std">
```
```suggestion
  <xacro:macro name="proximal_gazebo" params="prefix:=^ link_prefix:=^ prox_sensor:=^|'std'">
```

rhaschke replied:
As this is not in an eval context, i.e. `${}`, we don't need quotes to specify a simple string.

guihomework commented on sr_description/hand/xacro/full_hand.urdf.xacro:
```diff
@@ -1,81 +1,76 @@
+  <!-- eli (deprecated) is true for ellipsoid fingertip or false for standard finger -->
+  <!-- lf (deprecated) is true if we want a little finger, false otherwise -->
+  <xacro:macro name="shadowhand" params="hand_type:='hand_e' reflect prefix cable_mesh:=False fingers:=${[1,1,1,1,1]}
+          tip_sensors:=${5 * ['std']} mid_sensors:=${5 * ['std']} prox_sensors:=${5 * ['std']} palm_sensor:=std
```
```suggestion
          tip_sensors:=${5 * ['std']} mid_sensors:=${5 * ['std']} prox_sensors:=${5 * ['std']} palm_sensor:='std'
```

guihomework replied:
Can I then ask why `hand_type:='hand_e'` needs the quotes ? I find that with the quotes it is clearer

rhaschke replied:
None of these needs quotes. Thanks for pointing these out. I will push a cleanup commit to remove those everywhere.

guihomework commented on sr_description/hand/xacro/hand_lite.urdf.xacro:
```diff
@@ -7,19 +7,35 @@ xmlns:xacro="http://www.ros.org/wiki/xacro">
   <xacro:macro name="shadowhand"
-  params="muscletrans muscle bio bt_sp ubi eli reflect prefix extra_lite  cable_mesh:='false'">
+  params="hand_type:='hand_e' reflect prefix cable_mesh:=False fingers:=${[1,1,1,1]}
+          tip_sensors:=${4 * ['std']} mid_sensors:=${4 * ['std']} prox_sensors:=${4 * ['std']} palm_sensor:=std
```
```suggestion
          tip_sensors:=${4 * ['std']} mid_sensors:=${4 * ['std']} prox_sensors:=${4 * ['std']} palm_sensor:='std'
```

guihomework commented on sr_description/hand/xacro/one_finger_unit.urdf.xacro:
```diff
@@ -4,19 +4,29 @@ xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface"
-  <xacro:macro name="shadowhand"
-  params="muscletrans muscle bio bt_sp ubi eli reflect prefix">
+  <xacro:macro name="shadowhand" params="hand_type:='hand_e' reflect prefix cable_mesh:=False
+          tip_sensor:=std mid_sensor:=std prox_sensor:=std palm_sensor:=std
```
```suggestion
          tip_sensor:='std' mid_sensor:='std' prox_sensor:='std' palm_sensor:='std'
```

guihomework commented on sr_description/hand/xacro/th_ff_rf_ellipsoid_hand.urdf.xacro:
```diff
@@ -18,59 +9,14 @@ xmlns:xacro="http://www.ros.org/wiki/xacro">
+    <!-- translate old parameters to new ones, eventually updating tip_sensors -->
+    <xacro:include filename="$(find sr_description)/hand/xacro/backward_compat.urdf.xacro" />
+    <!-- forward to generic macro -->
+    <xacro:include filename="$(find sr_description)/hand/xacro/full_hand.urdf.xacro" />
```
If I am not mistaken, full_hand.urdf.xacro already includes the backward compat. Isn't it weird to do 2 inclusion here ?

Additionally, those specific macros created in the past for specific hands that could not be instantiated with the generic macro, were deleted in #109 because they are only used by a single robots/*.xacro . there is no need to make these backward compatible IMHO. People out there who were including those very specific macros in their own file should upgrade (so it should break for them), and they are very few, as compared to the main full_hand.xacro macro that was used in far more hands, (so should not break for them). If they were using the robots/*.xacro then the files can use the new macro as in https://github.com/shadow-robot/sr_common/pull/109/commits/9ed52f1dced2a31b733d03f01e87146f4d0348cf#diff-3c1228053895eb1008febb3f202141b33d09e83fb896afb77aa9c6c4a6b48799

rhaschke replied:
We should not break existing code without a deprecation phase.

guihomework replied:
I accept you statement on the deprecation phase. 
You did not answer on the 2 inclusions when another file also uses the "forward" but without the 2 inclusions

rhaschke replied:
> If I am not mistaken, full_hand.urdf.xacro already includes the backward compat. Isn't it weird to do 2 inclusion here ?

As we want to [access tip_sensors a few lines below](https://github.com/ubi-agni/sr_common/blob/23c55bc11d0512a7934185e2d2c1350999d52ea6/sr_description/hand/xacro/th_ff_rf_ellipsoid_hand.urdf.xacro#L20-L21), we need to ensure that it is well-defined from the old-style params *before* we call the macro, which then would include the compat file again.

guihomework replied:
I understand the explanation now, that is just to provide the possibility to set the other fingers to any given sensors. I see now that such a case was possible with the macro before, but I had not anticipated this case since the robots/*.xacro matching this macro file uses 'std' for the 2 other fingers.

guihomework commented on sr_description/hand/xacro/thumb/thumb.urdf.xacro:
```diff
@@ -4,34 +4,27 @@
-  params="muscletrans bio bt_sp ubi eli reflect prefix parent is_lite">
+  params=" tip_sensor:=^|std mid_sensor:=^|std prox_sensor:=^|std reflect:=^ prefix:=^ parent is_lite">
+    <xacro:if value="${tip_sensor == 'bio'}">
+      <xacro:property name="mid_sensor" value="bio" />
```
:+1:  I had not seen this corner case where `tip_sensor` and `mid_sensor` have different sensors. In the finger only `tip_sensor` is considered indeed for middle and distal phalanges, but thumb was expecting `tip_sensor` and `mid_sensor` to be identical when using 'bio'.

guihomework commented:
> That's what the [compatibility test](https://github.com/ubi-agni/sr_common/blob/pr-modernize-modularize-xacro/sr_description/test/test_compatibility.py) compares as well. I'm not sure where these attributes got lost. @guihomework, maybe you have an idea?
> 


I think I traced the problem to the presence of the include of the full_hand (defining the macro shadowhand) within the definition of the macro shadowhand there https://github.com/shadow-robot/sr_common/pull/120/files#diff-d505b43845bf4e41bf9a24883f9c24aef30837e1f512253f29313889990ce805R14

I was first surprised a macro redefinition was possible, so I tried to rename one of the macro either the exterior, or the interior one, but it does not change anything as long the include is inside the outer macro redefinition
Of course with the same macro name, one cannot put the include outside the new definition. It fails with 
 invalid parameter "fingers" 
because the macro call inside, wants to call the newly defined macro (recursivity issue ahead), instead of the included one.

I find this already too tricky for me, so cannot provide a fix. I hope this pin pointing suffices to guide to a possible true cause (within xacro itself I suppose)

guihomework commented:
@rhaschke I think I am done looking at the code and answering your questions. Everything I found was bad understanding of the code. sorry for the noise.
the compat test is not python2.7 compatible so I cannot test it myself under melodic, even calling it with python3 and some of the rospkg for python3 we have fails on Bionic.
I suppose it does not matter, @beatrizleon and @devops-shadowrobot  will confirm it does what they need.

rhaschke committed:
Deprecate specific hand.urdf.xacro files

rhaschke commented:
> I think I traced the problem to the presence of the include of the full_hand within the definition of the macro shadowhand.

Thanks for tracing this down! I will have a look.

rhaschke commented:
Indeed, the culprit was an issue in xacro: namespaces imported within a macro were not lifted. Fixed via https://github.com/ros/xacro/pull/285.

guihomework commented:
> Indeed, the culprit was an issue in xacro: namespaces imported within a macro were not lifted. Fixed via [ros/xacro#285](https://github.com/ros/xacro/pull/285).

I would never have generated such a case my self as I would never have attempt redefining a macro that has the same name that itself calls the macro name.
This "courage" of doing so requires a great knowledge on how xacro effectively works to know it will do what is desired.

rhaschke commented:
The problem was actually more fundamental: Any namespaces imported by a `xacro:include` within a macro were ignored - even if the include file doesn't do anything else than providing those namespaces. See the corresponding unit test I added in xacro for an example.
Redefining macros (or properties) is perfectly fine in xacro. As macros create a new scope, all symbols newly defined within the macro don't even affect the parent scope. That is, the original macro stays untouched when doing so.

rhaschke commented:
> I think I am done looking at the code and answering your questions.

Thanks a lot. Do you mind to also formally approve this PR if you are fine with it?

guihomework commented:
> > I think I am done looking at the code and answering your questions.
> 
> Thanks a lot. Do you mind to also formally approve this PR if you are fine with it?

I cannot test it easily immediately, I need to update my noetic/focal install first. I did not prepare my PR for noetic so cannot know what else can happen without testing.

guihomework commented:
> The problem was actually more fundamental: Any namespaces imported by a `xacro:include` within a macro were ignored - even if the include file doesn't do anything else than providing those namespaces. See the corresponding unit test I added in xacro for an example. Redefining macros (or properties) is perfectly fine in xacro. As macros create a new scope, all symbols newly defined within the macro don't even affect the parent scope. That is, the original macro stays untouched when doing so.

Not trivial to me, but probably fine for python developers.

rhaschke commented:
> I cannot test it easily immediately, I need to update my noetic/focal install first.

Note that the noetic-devel branches of this package as well as of `xacro` both work fine under Melodic as well. No need to install Noetic/Focal for testing.

beatrizleon commented:
@rhaschke  We were working on refactoring this repo with the help of @guihomework with the following purposes:
1. Be able to use one xacro to build all our hands (removing the need to create a xacro per hand configuration inside robots folder)
2. Have a way to easily specify the number of fingers and tactile sensors 
3. Introduce a way to specify the hand version differentiating the meshes on each of them.

I had a work in progress branch: https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes. It was based on the changes that @guihomework did on his proposed branch which were really helpful. With this new refactoring, I will base my branch on this one and show you what is the idea. It will be great to improve this as it has lots of deprecated code, unused meshes, etc 

Thanks both for you help with this. It is going slowly because of my lack of time to spend coding but hopefully with all your help, we will be able to finish this refactoring soon.

beatrizleon commented:
I checked the CI error and the issue is: 
`Copyright check failure: There are 1 files without copyright notices:
./sr_description/test/test_compatibility.py`

rhaschke commented:
> I had a work in progress branch: https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes, based on #109.
With this new refactoring, I will create a new branch based on this one and show you what is the idea. It will be great to improve this as it has lots of deprecated code, unused meshes, etc.

I will shortly rebase this PR onto the latest noetic-devel and apply the announced squashes.

rhaschke commented:
Rebase finished, which resolved the conflicts with upstream.
I kept the original, unsquashed version of this PR in https://github.com/ubi-agni/sr_common/tree/PR%23120-long.
One can easily validate the backward compatibility there by inspecting the [history of `test/robots.urdf`](https://github.com/ubi-agni/sr_common/commits/PR%23120-long/sr_description/test/robots.urdf)

> I checked the CI error and the issue is:
> `Copyright check failure: There are 1 files without copyright notices: ./sr_description/test/test_compatibility.py`

I added a copyright notice. Can you check why CI fails now, please?

guihomework commented:
> > I cannot test it easily immediately, I need to update my noetic/focal install first.
> 
> Note that the noetic-devel branches of this package as well as of `xacro` both work fine under Melodic as well. No need to install Noetic/Focal for testing.

well, I updated the branch of this PR #120 in my workspace, called a build + a 
`catkin build  --catkin-make-args run_tests -- sr_description`
and there no nosetest result
```
in ./test_results/sr_description/nosetests-test.xml
<?xml version="1.0" encoding="UTF-8"?><testsuite name="nosetests" tests="0" errors="0" failures="0" skip="0"></testsuite>
 (END)
```
if I call the test_compatibility.py manually, it fails to find subtest
```
ERROR: test_files (__main__.TestEquality)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_compatibility.py", line 153, in test_files
    with self.subTest(msg='Checking {}'.format(name)):
AttributeError: 'TestEquality' object has no attribute 'subTest'
```
if I understand well nosetest are based on unittest so I tried to check if I missed a dependency but rosdep did no indicate any because unittest is supposed to be part of python
https://github.com/ros/catkin/issues/735

my bionic is up-to-date, so I do not know what I am missing now. I therefore confirm that I cannot test fully on melodic

rhaschke commented:
@guihomework, thanks for this precise error description. With this I was able to file a fix: https://github.com/ubi-agni/sr_common/commit/5fd18833c1859b85d44548d32686d9080477096f, which is already squash-merged into the latest state of this PR. Please update again.

guihomework commented:
> @guihomework, thanks for this precise error description. With this I was able to file a fix: [ubi-agni@5fd1883](https://github.com/ubi-agni/sr_common/commit/5fd18833c1859b85d44548d32686d9080477096f), which is already squash-merged into the latest state of this PR. Please update again.

In melodic the nosetest still was empty after a `catkin build --catkin-make-args run_tests -- sr_description`
with verbose mode of catkin I could see the command and with verbosity level to 3 in nosetest I could find that the test_compatibility is skipped due to being executable
```
/vol/famula/nightly/bin/nosetests-2.7 -P --process-timeout=60 --where=/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test --with-xunit --xunit-file=/home/myname/work/dev/catkin_ws_shadow/build/sr_description/test_results/sr_description/nosetests-test.xml --verbosity=3
nose.config: INFO: Set working dir to /home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.selector: INFO: /home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_arms.sh is executable; skipped
nose.selector: INFO: /home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py is executable; skipped
nose.selector: INFO: /home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_hands.sh is executable; skipped
```

once file node set to non-executable, the nosetest runs but the failures are not clear, even with option -d
```
======================================================================
FAIL: test_files (test_compatibility.TestEquality)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 198, in test_files
    xml_matches(process(old_file), process(new_file), ignore_nodes=[xml.dom.Node.COMMENT_NODE])
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 171, in xml_matches
    return xml_matches(a.documentElement, b, ignore_nodes)
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 173, in xml_matches
    return xml_matches(a, b.documentElement, ignore_nodes)
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 175, in xml_matches
    return nodes_match(a, b, ignore_nodes)
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 138, in nodes_match
    if not all_attributes_match(a, b):
  File "/home/myname/work/dev/catkin_ws_shadow/src/shadow_robot/sr_common/sr_description/test/test_compatibility.py", line 95, in all_attributes_match
    raise AssertionError('Different number of attributes')
AssertionError: Different number of attributes


one does not know which file has differences

rhaschke commented:
> I had a work in progress branch: https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes, based on #109.
> With this new refactoring, I will create a new branch based on this one and show you what is the idea.

@beatrizleon, I shortly looked into https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes, but I'm not sure I did get all your goals. What I understood is:
- [ ] Rename hand, palm, and forearm files following a common scheme indicating the version with a single-char suffix (`e`, `g`, `m`)
  I like this idea. But, may I suggest using longer names. For me, the distinction doesn't become clear from these short ones.
  Could you please, provide a list of all to-be-distinguished versions and their long descriptions? This would help enormously.
- [ ] Restructure and rename mesh files into a clean folder structure, moving unused ones to `archive`.
  Again, generally a very good idea. Please keep more verbose names, though.
  Why keeping files in `archive` and not delete them? git has the history, hasn't it?
- [ ] Fix inaccuracies in model parameters (fca9513ce7e95d5a172c0e5fbb7d6d988692f724)
- [ ] Simplify parameter specification (5ee25811da86250243ed2081196fdb16b7a5f59b)
  I strongly agree that the lists need to be improved. I worked on some extensions to `xacro` to facilitate this.
  Please review https://github.com/ubi-agni/sr_common/pull/3, which is my suggestionn how to handle the specification of `fingers`.
  The same approach can be applied to the other parameters. My main concern with the current `list` approach is that the order of fingers is not obvious. Before continuing implementing anything along this line, we should agree on the way to go. My suggestion is as follows:
  - Use "dotified" dicts internally, i.e. `tip_sensors[2]` becomes `tip_sensors.mf` etc.
  - Specifying parameters as dicts is possible, but cumbersome as pointed out by @guihomework - and I agree:
    `tip_sensors = ${xacro.dotify(dict(th='bio', ff='bio', mf='std', rf='std', lf='none'))}`
    What about textual alternatives (as you suggest in 5ee25811da86250243ed2081196fdb16b7a5f59b) like these:
     - [ ] `tip_sensors = bio` for BioTacs on all fingers
     - [ ] `tip_sensors = 'bio, bio, std, std, none'` for different sensors on different fingers (association still not obvious)
     - [ ] `tip_sensors = 'th=bio, ff=bio, mf=std, rf=std, lf=none'` (same using dict syntax)

     Obviously, it requires some calculation to translate these text expressions into dicts. I can propose corresponding code as soon as wee agreed on the way to go. With my xacro extensions it will also be possible to validate those inputs and warn/error on any issues.

guihomework commented:
I am not sure a new point of view was explicitly requested again on the parameter specification, but I can remind that in URDF several "vectorized" parameters are passed by space separated strings (xyz, rpy) without dict. So I would understand a space separated string for all fingers / sensors of the fingers, and have an implicit order for all of them.

beatrizleon commented:
> > I had a work in progress branch: https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes, based on #109.
> > With this new refactoring, I will create a new branch based on this one and show you what is the idea.
> 
> @beatrizleon, I shortly looked into https://github.com/shadow-robot/sr_common/tree/F_refactoring_meshes, but I'm not sure I did get all your goals. What I understood is:
> 
> * [ ]  Rename hand, palm, and forearm files following a common scheme indicating the version with a single-char suffix (`e`, `g`, `m`)
>   I like this idea. But, may I suggest using longer names. For me, the distinction doesn't become clear from these short ones.
>   Could you please, provide a list of all to-be-distinguished versions and their long descriptions? This would help enormously.

These are the different hand types of hands we have:
Hand E: This is the dexterous hand with 5 fingers
Hand G: This is the hand lite version. It can be lite (4 fingers) or extra lite (3 fingers). It has no wrist dof.
Hand M: This is the muscle hand. We haven't produce them for a long time. Would you need to keep compatibility with them?

Hand plus is just a newer version of hand E so we will remove it from the hand types.

> * [ ]  Restructure and rename mesh files into a clean folder structure, moving unused ones to `archive`.
>   Again, generally a very good idea. Please keep more verbose names, though.
>   Why keeping files in `archive` and not delete them? git has the history, hasn't it?

Definitely, I was planned to delete it once I finished the refactoring. 

> * [ ]  Fix inaccuracies in model parameters ([fca9513](https://github.com/shadow-robot/sr_common/commit/fca9513ce7e95d5a172c0e5fbb7d6d988692f724))

Yes, fixing inaccuracies and in general we have created a more accurate meshes of some links. I will commit them soon.

> * [ ]  Simplify parameter specification ([5ee2581](https://github.com/shadow-robot/sr_common/commit/5ee25811da86250243ed2081196fdb16b7a5f59b))
>   I strongly agree that the lists need to be improved. I worked on some extensions to `xacro` to facilitate this.
>   Please review [Improve readability ubi-agni/sr_common#3](https://github.com/ubi-agni/sr_common/pull/3), which is my suggestionn how to handle the specification of `fingers`.
>   The same approach can be applied to the other parameters. My main concern with the current `list` approach is that the order of fingers is not obvious. Before continuing implementing anything along this line, we should agree on the way to go. My suggestion is as follows:
>   
>   * Use "dotified" dicts internally, i.e. `tip_sensors[2]` becomes `tip_sensors.mf` etc.
>   * Specifying parameters as dicts is possible, but cumbersome as pointed out by @guihomework - and I agree:
>     `tip_sensors=${xacro.dotify(dict(th='bio', ff='bio', mf='std', rf='std', lf='none'))}`
>     What about textual alternatives (as you suggest in [5ee2581](https://github.com/shadow-robot/sr_common/commit/5ee25811da86250243ed2081196fdb16b7a5f59b)) like these:
>     
>     * [ ]  `tip_sensors=bio` for BioTacs on all fingers
>     * [ ]  `tip_sensors='bio, bio, std, std, none'` for different sensors on different fingers (association not obvious)
>     * [ ]  `tip_sensors='th=bio, ff=bio, mf=std, rf=std, lf=none'` (same using dict syntax)
>     
>     Obviously, it requires some calculation to translate these text expressions into dicts. I can propose corresponding code as soon as we agreed on the way to go. With my xacro extensions, it will also be possible to validate those inputs and warn/error on any issues.

This is very important for us at the moment. We have developed an auto-detection tool that removes the need to specify the hand_serial and hand_id on the launch file (e.g. https://github.com/shadow-robot/sr-config/blob/demohand_A/sr_ethercat_hand_config/launch/sr_rhand.launch) detecting them at run time per each hand. The idea is that we will deprecate the need to use sr_config branch per hand. Instead we want to have a new repo that will have a folder per hand with the specific calibration and a configuration file which will contain:
* side
* hand type
* hand version
* number of fingers
* tactile sensors

The idea is that with those we could use a xacro to autogenerate the correct urdf. Similar to what I was trying to do here: https://github.com/shadow-robot/sr_common/blob/F_refactoring_meshes/sr_description/robots/sr_hand_e.urdf.xacro
But as you noticed, I was having a hard time passing them from sr_interface.

I agree with you that the list is not obvious and it leads to confusion. From your suggestions, I would prefer: `tip_sensors='th=bio, ff=bio, mf=std, rf=std, lf=none'` (same using dict syntax)

rhaschke committed:
Make fingers parameter a string

guihomework commented:
@beatrizleon so if I understand well the auto-generation part, you want the hand driver to find out about the hands on the ethercat bus, then look up in a matching folder (probably per hand id since they are unique), and from a file inside the folder,  generate the urdf.
May I suggest to use one yaml.file that the xacro can read easily. according to the hand id, it can directly read the file and fetch the required parameters. We do this here https://github.com/ubi-agni/agni_robots/blob/melodic-devel/agni_description/robots/robots.yaml and https://github.com/ubi-agni/agni_robots/blob/melodic-devel/agni_description/robots/shadow_hand_ati.urdf.xacro#L7

rhaschke commented:
> These are the different hand types of hands we have: 
> * Hand E: This is the dexterous hand with 5 fingers
> * Hand G: This is the hand lite version. It can be lite (4 fingers) or extra lite (3 fingers). It has no wrist dof. 
> * Hand M: This is the muscle hand. We haven't produce them for a long time.
> Hand plus is just a newer version of hand E so we will remove it from the hand types.

What about `hand_mt`? What kind of variant is that one?

Are these official product names? If so, we should of course stick with them.
What about these `hand_type` specifiers: 
* `E` with synonyms: `hand_e` (for compatiblity), `motor` (the one most easiest to remember for me :wink:)
* `E+` with synonyms: `hand_e_plus`, `motor+` (if this is needed at all, i.e. has different dimensions than `E`)
* `G` with synonyms: `lite`
* `G-` with synonyms: `extra_lite`
* `M` with synonyms: `muscle`

We can easily translate these synonyms into the standard form _once_, such that the code only needs to handle the standard form.

> The idea is that with those we could use a xacro to autogenerate the correct urdf. Similar to what I was trying to do here: https://github.com/shadow-robot/sr_common/blob/F_refactoring_meshes/sr_description/robots/sr_hand_e.urdf.xacro.

This is perfectly possible.

> I agree with you that the list is not obvious and it leads to confusion. From your suggestions, I would prefer: `tip_sensors='th=bio, ff=bio, mf=std, rf=std, lf=none'` (same using dict syntax)

Good. I can propose corresponding code. Using a YAML file to read all required parameters is definitely a good idea!

rhaschke commented:
> I agree with you that the list is not obvious and it leads to confusion. From your suggestions, I would prefer: `tip_sensors='th=bio, ff=bio, mf=std, rf=std, lf=none'` (same using dict syntax)

I implemented the basic scheme for this (and all the other sensor specifications) in https://github.com/ubi-agni/sr_common/pull/3 / https://github.com/ubi-agni/sr_common/pull/3/commits/56d4718abf5f8ed0cda4932058f530990aaaa9ac. There are still some corner cases and I tested on Noetic only, but one can get an idea of the approach. As stated before, this strongly depends on [our xacro fork](https://github.com/ubi-agni/xacro).

rhaschke committed:
Apply new sensor specification

rhaschke committed:
Internally use dotified dict for *_sensors

rhaschke committed:
Python 2.x compatibility

As xacro returns unicode objects from property evaluations
we cannot apply str.lower to them in Python 2.x
In Python 3.x str and unicode is essentially the same.

guihomework commented:
@rhaschke I am sorry there was a misunderstanding that delayed me testing again. After I reported https://github.com/shadow-robot/sr_common/pull/120#issuecomment-932933525 I was waiting for feedback on the error and did not retry after the force-push you did. 
I re-tested and the assertion is much better now including a copy of the failing files in tmp. Thanks for the fix.

I see no error generated when using the xacro fork of ubi

rhaschke committed:
Remove ubi params

rhaschke committed:
Remove use of ubi param

rhaschke committed:
Simplify if-else clauses

Improve extensibility (adding more sensors) by emloying an exclusive-if-and-final-else
strategy, utilizing a helper variable. This way we avoid augmenting the list in the
'sensor in [...]' tests.

rhaschke commented:
I factored out the ubi tactile sensor definitions.

rhaschke commented on sr_description/hand/xacro/thumb/thmiddle.gazebo.xacro:
```diff
@@ -1,25 +1,25 @@
+            <bumperTopicName>contacts/${prefix}th/middle</bumperTopicName>
+          </plugin>
+        </sensor>
+        <selfCollide>true</selfCollide>
```
@guihomework: This `selfCollide` tag is part of the `unless` clause here, while in all other places it's outside.
Is this one wrong here?

guihomework replied:
yes it makes no sense to remove the selfcollide here and not at other places, yet another mistake of mine. 

side note : I am not convinced anymore by the current set of true/false on the self collide in the fingers, but I think it is a tricky matter (at least I remember it was not obvious) and if the current set works one should not try change it.

guihomework commented:
> I factored out the ubi tactile sensor definitions.

I don't know why this is necessary, since shadow robot company supports the ubi sensors, even in the driver.

rhaschke commented:
> I don't know why this is necessary, since shadow robot company supports the ubi sensors, even in the driver.

As you said yourself: they can only be used with our custom urdf repos, which will never merged into public.

guihomework commented:
> > I don't know why this is necessary, since shadow robot company supports the ubi sensors, even in the driver.
> 
> As you said yourself: they can only be used with our custom urdf repos, which will never merged into public.

Not exactly what I said : I said that the taxels cannot be used without our tactile_toolbox, and could maybe be removed, but the visual/collision of the ubi could stay there as they do not need the tactile_toolbox. They were merged by shadow there https://github.com/shadow-robot/sr-ros-interface/pull/253  and there was a distinction in the urdf for ubi that I had added already at https://github.com/shadow-robot/sr-ros-interface/pull/147 and was also accepted upstream.

We always wanted to have as maximum things merged upstream as possible. If this  an effort to now clean up their repo from tip sensor they do not sell, sure, then this the reason, but it is not a problem to have our visual tips even without tactile_toolbox.

rhaschke commented:
As a maintainer, I wouldn't like to have external sources in my repo that I cannot validate.
From my point of view, it is cleaner that we maintain our sensor addons on our side.
Yes, my point of view has changed on this since 2014 - due to much more experience in maintaining repositories.

guihomework commented:
@beatrizleon in place of the team in Paris UPMC who has a ShadowArm and won't probably notice those discussions here, I would say, either they should fork to keep it (please announce it to the customers), or your company should create a legacy sr_description package, with things that you supported in the past, but won't anymore. This way it stays visible (and not hidden in history that is no easy to find), in the state they are now.
I personally like companies who still provide old drivers for winXP etc in a legacy section of their website. It is just a matter of dragging the files along but they require less maintenance and are separated for the supported ones.

Regarding the muscle hands in my current team, I think they were never used with ROS, but the same about the meshes and URDFs as for the Shadow Muscle arm applies IMHO. If for any reason someone would like to integrate the muscles hands in ROS on his own, finding the resources without digging in history would be nice.

rhaschke commented:
Dear @beatrizleon, skimming through your changes in 417b7a0b4daebc4a0a0ca670703c8fe84488469e, I noticed the following potential issues:
- It seems like you removed the `std` sensor type. This was intended to be used as the default for Gazebo.
- You renamed the hand_type `motor` to `E` in `process_hand_type_parameter.xacro`, but not yet in all other locations.
- It looks like you moved many mesh files into an `archive` subfolder. However, git didn't recognized that they were renamed.

In general, it was rather hard to review your single commit. It would be helpful if you could split commits into smaller units with a commit message clearly stating what has changed.
